docs/configuration-file.md
# Configuration File

You can configure the extension's behavior by adding a `tt3.config.json` file to the top folder of your game. If you don't include one, the extension uses the following default configuration:

```json
{
    "build": {
        "storySourceFiles": [
            "src/**/*.{tw,twee}",
            "src/**/*.css",
            "src/**/*.js",
            "src/**/*.{otf,ttf,woff,woff2}",
            "src/**/*.{gif,jpeg,jpg,png,svg,tif,tiff,webp}",
            "src/**/*.{aac,flac,m4a,mp3,oga,ogg,opus,wav,wave,weba}",
            "src/**/*.{mp4,ogb,webm}",
            "src/**/*.vtt"
        ],
        "includeSourcePaths": ["include"],
        "outputPath": "build",
        "storyFormatPaths": [".storyformats"]
    },
    "tt3": {
        "disabledDiagnostics": []
    }
}
```

## Full Contents

```json
{
    "build": {
        "storySourceFiles": [
            "src/**/*.{tw,twee}",
            "src/**/*.css",
            "src/**/*.js",
            "src/**/*.{otf,ttf,woff,woff2}",
            "src/**/*.{gif,jpeg,jpg,png,svg,tif,tiff,webp}",
            "src/**/*.{aac,flac,m4a,mp3,oga,ogg,opus,wav,wave,weba}",
            "src/**/*.{mp4,ogb,webm}",
            "src/**/*.vtt"
        ],
        "includeSourcePaths": ["include"],
        "ignores": ["src/**/skip.tw"],
        "outputPath": "build",
        "outputFilename": "game.html",
        "startPassage": "Start"
    },
    "tt3": {
        "disabledDiagnostics": []
    }
}
```

### `build` Section

Settings for building the story.

- `storySourceFiles`: Files (relative to the project's root) to be included in the built story. Supports [glob patterns](https://code.visualstudio.com/docs/editor/glob-patterns). Allows any Tweego-supported files except for `.tw2`, `.twee2`, `.htm`, and `.html`.
- `includeSourcePaths`: Paths (relative to the project's root) whose contents will be included as-is alongside the built story.
- `ignores` (Optional): Paths to files (relative to the project's root) to be excluded from storySourceFiles and includeSourcePaths. Supports [glob patterns](https://code.visualstudio.com/docs/editor/glob-patterns).
- `outputPath` (Optional): Path (relative to the project's root) to write the story file to. If omitted, defaults to `build`.
- `outputFilename` (Optional): Name for the story file. If omitted, defaults to the game's filename from the StoryTitle passage, with spaces replaced by hyphens.
- `storyFormatPaths`: Path to story formats. Story formats are contained in subdirectories whose name matches the pattern `<format>-<major>-<minor>-<patch>`. Path is relative to the project's root.
- `startPassage` (Optional): Name of the starting passage. Will override the one set in the game's `StoryData` passage.

### `tt3` Section

Settings for the Twine (Twee 3) extension.

- `disabledDiagnostics`: A list of diagnostic codes to be disabled project-wide. Can be populated by using a [Quick Fix action](https://code.visualstudio.com/docs/editing/refactoring#_code-actions-quick-fixes-and-refactorings) on an error in your game.


CHANGELOG.md
# Change Log

Changes to the extension.

## Unreleased

Three big changes:

- Story build configuration is now handled a `tt3.config.json` file instead of in extension settings.
- Diagnostics have been extensively re-worked.
    - The extension now reports JavaScript errors.
    - It flags variables that (it thinks) have never been defined.
    - Every diagnostic has a Quick Action to let you disable it for a single passage or globally (through the config file).
- Passage tag handling has been greatly improved.
    - Tag auto-complete? Yes, please!
    - Find every reference to a tag or rename a tag the same way you can a passage or variable.

### Added

- JavaScript diagnostics (warnings and errors) are now reported.
- Diagnostics can now be disabled in a passage by adding `tt3-disable` followed by comma-separated list of diagnostics to the passage's tags: `[tt3-disable <diagnostic-code-1>,<diagnostic-code-2>]`.
- Added Quick Fix actions for disabling diagnostics for a passage or globally.
- You can now search for and rename passage tags the same way you can passages and variables.
- Passage tags now auto-complete.
- Variables and properties that haven't been defined now create a warning across both story format code (Chapbook/SugarCube) and JavaScript sections.
- Now auto-completes object properties from expressions like `var['prop'] = { prop1: 2 }`.
- SugarCube: Variables and properties in script passages and `<<script>>` macros are now properly indexed, including `State.variables.var` and `state.temporary.temp` variables.
- SugarCube: Special variables `_args` and `_contents` now produce warnings if used outside of `<<widget>>` macros.
- SugarCube: Macros inside link setters (such as `[[passage][<<set $var to 1>>]]`) now generate a warning about them not being evaluated.
- Chapbook: Added parsing variables and properties in script passages and `[JavaScript]` modifier blocks, including variables set in calls to `engine.state.set()`.

### Removed

- Removed extension options to enable or disable warnings on unknown macro and unknown passage references, as all diagnostics can now be disabled on a project or per-passage basis.

### Changed

- The extension now offers to set up the workspace's directories when the extension starts instead of waiting until a `.twee` file is opened.

### Fixed

- When running a game in VS Code, error modals are now shown in VS Code instead of being silently swallowed.
- SugarCube: Improved syntax highlighting of twee links.
- SugarCube: Fixed parsing of special variables in `<<script>>` macros.
- SugarCube: No longer incorrectly indexes apparent variables in non-data HTML attributes.

## [1.1.5] - 2026 04 12

### Fixed

- Build system no longer percent-encodes spaces in included files' filenames. (issue #7)
- Spaces are now allowed in local image paths when running games.
- SugarCube: `[img]` markup images now show properly in running games. (issue #9)
- Chapbook: Inserts with a quote mark in the text between them (such as `{var1} ' {var2}`) no longer ignore the second insert.

## [1.1.4] - 2025 06 25

### Added

- Chapbook: Variable section contents are now decorated in the editor to make them more visually distinctive when editing them.

### Fixed

- Properly loads local images when running games. (issue #5)
- Chapbook: Modifier completions now work as expected when the cursor is at a semicolon.
- SugarCube: Now properly recognizes the `<<repeat>>` macro. (issue #6)
- SugarCube: Warns if `<<do>>` macro is used before SugarCube 2.37.0.

## [1.1.3] - 2025 03 18

A collection of Chapbook enhancements/fixes, as well as bringing SugarCube support up to parity with [twee3-language-tools](https://github.com/cyrusfirheir/twee3-language-tools/) v0.32.0.

### Changed

- Chapbook: Insert completions aren't suggested after `{` in Javascript-modified sections.
- Chapbook: Variables are no longer auto-suggested in strings inside the vars section.
- SugarCube: Whitespace is now accepted after a `|` macro parameter variant. (twee3-language-tools issue [#176](https://github.com/cyrusfirheir/twee3-language-tools/issues/176))

### Fixed

- Build system can now build games whose source code is at the root level of the project.
- Fixed bug that caused issues if any files had no extension.
- Chapbook and SugarCube now properly handle `[script]` passages.
- Chapbook: Variables that start with `$` or `_` are now properly parsed.
- Chapbook: Removed spurious errors for a range of variable names when set in the vars section.
- Chapbook: `random.d4` and similar are no longer flagged as unrecognized variables.

## [1.1.2] - 2025 01 12

### Changed

- Reworked parsing routines to improve performance on large projects.

### Fixed

- Source code at the root level of a project is now indexed properly.
- SugarCube: Lexing errors are now logged against the correct text.
- SugarCube: Macros like `<<link>>` that accept link markup no longer cause spurious errors.
- SugarCube: Macros like `<<for>>` that take array or object containers are now parsed properly.
- SugarCube: `<<case>>` macros now accept `undefined` as a value.

## [1.1.1] - 2024 12 21

### Added

- Chapbook: Added folding ranges for modifiers.
- Chapbook: Modifier contents are now decorated in the editor to make them more visually distinctive when editing them.

### Changed

- Chapbook: Variable and property auto-completions now only include variables and properties set in a vars section.
- Chapbook: Modifiers now have a different syntax highlighting color than inserts so they're easier to distinguish.

### Fixed

- "Indexing Twine Project..." status bar item no longer hides the "Run Twine Game" item when the former appears.
- Chapbook: Fixed bug where renaming variables and properties didn't rename all variables or properties.
- Chapbook: Completions for modifiers after a semicolon no longer have an extra space.

## [1.1.0] - 2024 12 01

### Added

- The extension now shows on the status bar when the project is being indexed.

client/src/config.ts
import { z } from "zod";

import { DiagnosticCode, diagnosticCodeSet } from "@tt3/shared";

/**
 * Config file name.
 */
export const ConfigFilename = "tt3.config.json";

/**
 * Regex to match relative paths.
 */
const RelativePathRegex = /^(?!\/)(?![A-Za-z]:[\\/])(?!https?:\/\/).+/;

export const SupportedStoryFileTypes = [
    "*.{tw,twee}",
    "*.css",
    "*.js",
    "*.{otf,ttf,woff,woff2}",
    "*.{gif,jpeg,jpg,png,svg,tif,tiff,webp}",
    "*.{aac,flac,m4a,mp3,oga,ogg,opus,wav,wave,weba}",
    "*.{mp4,ogb,webm}",
    "*.vtt",
];

const RelativePath = z.string().regex(RelativePathRegex);
const RelativePathArray = z.array(RelativePath);

export const Tt3ConfigSchema = z
    .object({
        build: z
            .object({
                // The storySourceFiles description gets a markdown version in `scripts/build-schema.ts` and must be synchronized
                storySourceFiles: RelativePathArray.min(1)
                    .default(
                        SupportedStoryFileTypes.map((type) => `src/**/${type}`),
                    )
                    .describe(
                        "Files (relative to the project's root) to be included in the built story. Supports glob patterns. Allows any Tweego-supported files except for .tw2, .twee2, .htm, and .html.",
                    ),
                includeSourcePaths: RelativePathArray.default([
                    "include",
                ]).describe(
                    "Paths (relative to the project's root) whose contents will be included as-is alongside the built story.",
                ),
                ignores: RelativePathArray.optional().describe(
                    "Paths to files (relative to the project's root) to be excluded from storySourceFiles and includeSourcePaths. Supports glob patterns.",
                ),
                // The outputPath description gets a markdown version in `scripts/build-schema.ts` and must be synchronized
                outputPath: RelativePath.default("build").describe(
                    "Path (relative to the project's root) to write the story file to. If omitted, defaults to 'build'.",
                ),
                outputFilename: z
                    .string()
                    .optional()
                    .describe(
                        "Name for the story file. If omitted, defaults to the game's filename from the StoryTitle passage, with spaces replaced by hyphens.",
                    ),
                // The storyFormatPathx description gets a markdown version in `scripts/build-schema.ts` and must be synchronized
                storyFormatPaths: RelativePathArray.default([
                    ".storyformats",
                ]).describe(
                    "Path to story formats directory. Story formats are contained in subdirectories whose name matches the pattern <format>-<major>-<minor>-<patch>. Path is relative to the project's root.",
                ),
                startPassage: z
                    .string()
                    .optional()
                    .describe(
                        "Name of the starting passage. Will override the one set in the game's StoryData passage.",
                    ),
            })
            .prefault({}),
        tt3: z
            .object({
                disabledDiagnostics: z.array(z.string()).default([]),
            })
            .prefault({})
            .describe("List of diagnostic codes to be disabled project-wide."),
    })
    .prefault({});
type Tt3Config = z.infer<typeof Tt3ConfigSchema>;

/**
 * Current TT3 project configuration.
 */
export let currentConfig: Readonly<Tt3Config> = Tt3ConfigSchema.parse({});

/**
 * Update the project's config wholesale.
 *
 * @param newConfig New configuration.
 * @returns Description of any errors found in the config, or undefined if none.
 */
export function updateConfig(newConfig: Tt3Config): string | undefined {
    const maybeConfig = Tt3ConfigSchema.safeParse(newConfig);
    if (maybeConfig.success) {
        currentConfig = maybeConfig.data;
        const badCodes = validateConfigDiagnosticCodes();
        if (badCodes !== undefined) {
            return `Non-existent diagnostic codes: ${badCodes.join(", ")}`;
        }
    } else {
        return z.prettifyError(maybeConfig.error);
    }
}

/**
 * Update the extension's config from JSON.
 *
 * @param contents JSON contents of the config file.
 * @returns Description of any errors found in the config, or undefined if none.
 */
export function updateConfigFromJson(contents: string): string | undefined {
    // If contents is blank, turn it into a blank JSON object
    if (!contents) contents = "{}";

    let jsonContents;
    try {
        jsonContents = JSON.parse(contents);
    } catch (e) {
        return `The contents aren't valid JSON: ${(e as Error).message}`;
    }
    return updateConfig(jsonContents);
}

/**
 * Check that the current configuration's diagnostic codes are all real ones.
 *
 * @returns List of unrecognized diagnostic codes, or undefined if none.
 */
function validateConfigDiagnosticCodes(): string[] | undefined {
    // TODO When we target es2024 (starting w/VS Code 1.123 2026-06-03), use Set operations
    const badCodes = currentConfig.tt3.disabledDiagnostics.filter(
        (c) => !diagnosticCodeSet.has(c as DiagnosticCode),
    );
    currentConfig.tt3.disabledDiagnostics =
        currentConfig.tt3.disabledDiagnostics.filter((c) =>
            diagnosticCodeSet.has(c as DiagnosticCode),
        );
    if (badCodes.length > 0) return badCodes;
}
